### PR TITLE
Prototype `mir_verify` command

### PR DESCRIPTION
**Pull request by RyanGlScott**
This implements just enough scaffolding to support the basics of a `mir_verify` command and related scaffolding, which is one of the main goals of #1859. When I say "basic", I really do mean that: there are quite a few things that do not work currently. These include:

* `mir_alloc`/`mir_points_to` do not work yet.
* Overrides (unsafe or otherwise) do not work yet.
* There is no way to declare variables with `struct` or `enum` types.
* Likely other things.

These limitations notwithstanding, it is now possible to write MIR specifications for very simple functions such as the ones in the `test_mir_verify_basic` integration test. I will be adding additional capabilities in subsequent patches.

Most of the code in this patch is not terribly exciting, as lots of it is cargo-culted from the LLVM and JVM backends. Ideally, we would find a way to deduplicate a lot more of this code, but this will not be straightforward without substantial refactoring in SAW. See #379.

In addition to the code itself, I have also expanded the SAW manual to mention the variety of new MIR-related commands added in this patch. Of particular interest is that `mir_verify` allows you to specify function identifiers in multiple ways, which is provided as a convenience to SAW users. See the manual for more details.

Checks off several boxes in #1859.

### CODE REVIEW

File: src/SAWScript/Crucible/MIR/Builtins.hs [hunk 1]
```diff
@@ -26,5 +210,952 @@ mir_load_module inputFile = do
...
+
+-- | Convert a 'Mir.FnSig' value to a 'FnSigView' value.
+fnSigView :: Mir.FnSig -> FnSigView
+fnSigView (Mir.FnSig argTys retTy abi spreadarg) =
+  FnSigView (map tyView argTys) (tyView retTy) abi spreadarg
+
+-- | Returns the Cryptol type of a MIR type, returning 'Nothing' if it is not
+-- easily expressible in Cryptol's type system or if it is not currently
+-- supported.
+cryptolTypeOfActual :: Mir.Ty -> Maybe Cryptol.Type
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/Builtins.hs:1009]: This describes how to convert a MIR type (`Mir.Ty`) to a Cryptol `Type`.

File: src/SAWScript/Crucible/MIR/Builtins.hs [hunk 2]
```diff
@@ -26,5 +210,952 @@ mir_load_module inputFile = do
...
+-- 'Mir.DefId'. Currently, the following types of function names are permittd:
+--
+-- * @<crate_name>/<disambiguator>::<function_name>: A fully disambiguated name.
+--
+-- * @<crate_name>::<function_name>: A name without a disambiguator. In this
+--   case, SAW will attempt to look up a disambiguator from the @crateDisambigs@
+--   map. If none can be found, or if there are multiple disambiguators for the
+--   given @<crate_name>@, then this will fail.
+findDefId :: Map Text (NonEmpty Text) -> Text -> TopLevel Mir.DefId
+findDefId crateDisambigs fnName = do
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/Builtins.hs:1055]: This function implements the MIR identifier lookup logic described in the SAW manual.

File: src/SAWScript/Crucible/MIR/Override.hs [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+         nameEnv = MS.csTypeNames spec
+     liftIO $ typeOfSetupValue cc tyenv nameEnv sval
+
+valueToSC ::
+  Sym ->
+  MS.ConditionMetadata ->
+  OverrideFailureReason MIR ->
+  Cryptol.TValue ->
+  MIRVal ->
+  OverrideMatcher MIR w Term
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/Override.hs:444]: This describes how to convert a `MIRVal` to a SAWCore `Term`.

**Review by RyanGlScott — COMMENTED** [review]
There are a lot of "plumbing-related" changes in this PR, which makes this a bit of a slog to review. I've made an effort to highlight what I think are the most interesting changes when compared to the existing LLVM/JVM backends.

File: src/SAWScript/Crucible/MIR/TypeShape.hs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        TupleShape ty [refTy, usizeTy]
+            (Empty
+                :> ReqField (RefShape refTy (M.TyUint M.B8) mutbl (BVRepr (knownNat @8)))
+                :> ReqField (PrimShape usizeTy BaseUsizeRepr))
+      where
+        -- We use a ref (of the same mutability as `ty`) when possible, to
+        -- avoid unnecessary clobbering.
+        refTy = case ty of
+            M.TyRef _ _ -> M.TyRef ty' mutbl
+            _ -> M.TyRef ty' mutbl
```
Comment (bboston7) [inline src/SAWScript/Crucible/MIR/TypeShape.hs:178]: The result of these cases is the same. Is that intentional?

File: src/SAWScript/Interpreter.hs [hunk 5]
```diff
@@ -3836,11 +3837,212 @@ primitives = Map.fromList
...
+    , "After `mir_execute_func`, it states that the function being"
+    , "verified is expected to perform the allocation."
+    ]
+
+  , prim "mir_assert" "Term -> MIRSetup ()"
+    (pureVal mir_assert)
+    Experimental
+    [ "State that the given predicate must hold.  Acts as `mir_precond`"
+    , "or `mir_postcond` depending on the phase of specification in which"
+    , "it appears (i.e., before or after `mir_execute_func`."
```
Comment (bboston7) [inline src/SAWScript/Interpreter.hs:3865]:
```suggestion
    , "it appears (i.e., before or after `mir_execute_func`)."
```

**Review by bboston7 — APPROVED** [review]
This broadly looks great! Thanks for pointing out places to focus more on. All of my comments are minor.

Also, Github won't let me put comments far away from your changes, so I'm putting one here: 
 On line 2359 in the manual the part that says "(or the corresponding JVM statements)" should say something like "(or the corresponding JVM or MIR statements)"

File: src/SAWScript/Crucible/MIR/Builtins.hs [hunk 6]
```diff
@@ -26,5 +210,952 @@ mir_load_module inputFile = do
...
+  [Crucible.GenericExecutionFeature Sym] ->
+  MethodSpec ->
+  [(a, MIRVal)] ->
+  [Crucible.LabeledPred Term AssumptionReason] ->
+  W4.ProgramLoc ->
+  [Lemma] ->
+  Crucible.SymGlobalState Sym ->
+  Bool {- ^ path sat checking -} ->
+  IORef MetadataMap {- ^ metadata map -} ->
+  IO (Maybe (Mir.Ty, MIRVal), Crucible.SymGlobalState Sym)
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/SAWScript/Crucible/MIR/Builtins.hs:727]: Hah, I was lazy when I copy-pasted this code from the LLVM/JVM backends, which also lack Haddocks for this function. (Not that that excuses my laziness, but that's how it happened.) I've added a separate commit that adds some Haddocks for the `verifySimulate` functions in all of the SAW backends.

File: src/SAWScript/Crucible/MIR/ResolveSetupValue.hs [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+              ]
+
+toMIRType ::
+  Cryptol.TValue ->
+  Either ToMIRTypeErr Mir.Ty
+toMIRType tp =
+  case tp of
+    Cryptol.TVBit -> Right Mir.TyBool
+    Cryptol.TVInteger -> Left (NotYetSupported "integer")
+    Cryptol.TVIntMod _ -> Left (NotYetSupported "integer (mod n)")
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/SAWScript/Crucible/MIR/ResolveSetupValue.hs]: Indeed, I'm not sure what `Z q` would mean for MIR, now that I think about it more closely. I've changed this from `NotYetSupported` to `Impossible`.

File: src/SAWScript/Crucible/MIR/TypeShape.hs (hunk 4 above)
Comment (RyanGlScott) [reply src/SAWScript/Crucible/MIR/TypeShape.hs]: While I didn't write this code (it was moved from a different module in `crucible-mir-comp`), I agree that this code is rather silly as written. I've pushed a separate commit to remove the unnecessary `case` expression.

File: src/SAWScript/Interpreter.hs (hunk 5 above)
Comment (RyanGlScott) [reply src/SAWScript/Interpreter.hs]: Well spotted. What's more, this same mistake is present in the docstrings for `llvm_assert` and `jvm_assert`, from where I originally copied this! I've pushed a separate commit which fixes all three copies of the mistake.

**Pull request merged**